lychee-lib/src/chain/mod.rs

```diff
@@ -0,0 +1,91 @@
+use core::fmt::Debug;
+
+use crate::Status;
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum ChainResult<T, R> {
+    Chained(T),
+    EarlyExit(R),
+}
+
+pub(crate) type RequestChain = Chain<reqwest::Request, Status>;
+
+#[derive(Debug)]
+pub struct Chain<T, R>(Vec<Box<dyn Chainable<T, R> + Send>>);
+
+impl<T, R> Default for Chain<T, R> {
+    fn default() -> Self {
+        Self(vec![])
+    }
+}
+
+impl<T, R> Chain<T, R> {
+    pub(crate) fn new(values: Vec<Box<dyn Chainable<T, R> + Send>>) -> Self {
+        Self(values)
+    }
+
+    pub(crate) fn append(&mut self, other: &mut Chain<T, R>) {
+        self.0.append(&mut other.0);
+    }
+
+    pub(crate) fn traverse(&mut self, mut input: T) -> ChainResult<T, R> {
+        use ChainResult::*;
+        for e in self.0.iter_mut() {
+            match e.handle(input) {
+                Chained(r) => input = r,
+                EarlyExit(r) => {
+                    return EarlyExit(r);
+                }
+            }
+        }
+
+        Chained(input)
+    }
+
+    pub(crate) fn push(&mut self, value: Box<dyn Chainable<T, R> + Send>) {
+        self.0.push(value);
+    }
+}
+
+pub(crate) trait Chainable<T, R>: Debug {
+    fn handle(&mut self, input: T) -> ChainResult<T, R>;
+}
+
+mod test {
+    use super::{ChainResult, ChainResult::*, Chainable};
+
+    #[derive(Debug)]
+    struct Add(i64);
```

Comment:
Nitpick: this could be a `usize`, since it won't get negative.
Same for Result below.